streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,638 @@
+            sample_ids (NDArray[np.int64]): The samples to download the shards of.
+        """
+        # Create or attach shard_states array.
+        name = f'{self._prefix}_shard_states'
+        size = len(self.shard_sizes) * np.uint8(0).nbytes
+        try:
+            shm = SharedMemory(name, True, size)
+        except:
+            shm = SharedMemory(name)
+        shard_states = np.ndarray(len(self.shard_sizes), buffer=shm.buf, dtype=np.uint8)
+
+        filename = os.path.join('/tmp', 'mds', self._prefix, '_shard_states_filelock')
+        shard_states_lock = FileLock(filename)
+
+        # Download loop.
+        num_samples = len(sample_ids)
+        while True:
+            # If we've started a new epoch early (__iter__ was called again), exit this thread
+            # because there can only be one epoch at once.
+            if epoch != self.next_epoch - 1:
+                break
+
+            # If we're out of samples this epoch, exit this thread because we are done downloading.
+            if self._download_index == num_samples:
+                break
+
+            # If we are requested to only pre-download so many samples, if we have as many or more
+            # downloaded already, we wait and check again later.
+            if self.predownload is not None:
+                samples_ahead = self._download_index - self._iter_index
+                if self.predownload <= samples_ahead:
+                    sleep(0.07)
+                    continue
+
+            # Download and decompress the shard for this sample, if not already done.
+            sample_id = sample_ids[self._download_index]
+            shard_id, _ = self.index.find_sample(sample_id)
+            self._download_or_await_shard(shard_states_lock, shard_states, shard_id)
+            self._download_index += 1
+
+    def __iter__(self) -> Iterator[Dict[str, Any]]:
+        """Iterate over all the samples in our partition.
+
+        Returns:
+            Iterator[Dict[str, Any]]: Each sample.
+        """
+        # Load resume state and create the new training session.
+        presumed_epoch = self.next_epoch
+        epoch, old_sessions = self._resume(presumed_epoch)
+        cur_session, _ = self._create_cur_session(epoch)
+
+        world = World()
+        assert world.num_workers == self._barrier.count
+        self._barrier()
+
+        # Update the pre-incremented epoch counter.
+        world = World()
+        if world.is_local_leader:
+            self.next_epoch = epoch + 1
+
+        self._barrier()
+
+        # Get the samples for this worker to process.
+        sessions = old_sessions + [cur_session]
+        sample_ids = self._get_partition(epoch, sessions, world)
+        if sample_ids is None:
+            return
+
+        # Iterate while downloading.
+        self._iter_index = 0
+        self._download_index = 0
+        Thread(target=self._download_thread, args=(epoch, sample_ids)).run()
```

Comment:
I am not very positive about this line w.r.t releasing the resources and killing the thread. I am thinking, can we modified it as below? 

1. assign a thread local variable and put `daemon=True` such as
```
self.thread = Thread(target=self._download_thread, args=(epoch, sample_ids))
self.thread.start()
```
2. And put everything under `try..finally` block with waiting and release the thread. For example

```
try:
  .......
  self.thread = Thread(target=self._download_thread, args=(epoch, sample_ids))
  self.thread.start()
  ......
finally:
  if self.thread is not None:
    self.thread.join()
    self.thread = None
```

Reply:
Setting daemon to True will let the main thread exit even though the workers are blocking. This will solves the hanging issue if any.

Reply:
redesigned, daemon=True